heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -92,9 +100,14 @@ private void runOnce() {
   }
 
   private void onExit() {
-    for (Runnable r : exitTasks) {
-      r.run();
+    int s = exitTasks.size();
+    for (int i = 0; i < s; i++) {
+      if (clearExitTasks) {
```

Comment:
is it guaranteed that the clear() function is called in the same thread? the concern is that if the clear() function is called in a different thread, it might be called after the if() check and you can still get the wrong index error. Since this object might be used by new code in future so there may not be a guarantee even if there s no problem for now. I think it might be better to have a comment to be clear that this flag is not thread safe and the clear functions are required to be called ONLY in the same thread.

Or I am wondering would it be safer to make onExit, executeTasksOnWakeup, clearTasksOnWakeup and  clearExitTasks synchronized?

Reply:
I have checked all code paths.  Ya I can confirm clear() is always called from the same thread.  I don't know how we can make nExit, executeTasksOnWakeup, clearTasksOnWakeup and clearExitTasks synchronized.  This would give rise to deadlock situations since tasks run in executeTasksOnWakeup() can in turn call clear().

Reply:
I was thinking of simply adding synchronized keyword to these functions. I believe Java synchronized blocks in Java are reentrant so it is not a deadlock.

Reply:
I see what you are saying, but should we add them? It gives a developer the sense that these operations are thread safe now.  Should we even allow that in the architecture.  I think that is a bigger discussion. I can't think of a use case for that now.  It also might not be wise to allow developers to call these from potentially other threads since heron's instance architecture is clearly defined to be 2 thread based, gateway and slave.